nova/virt/vmwareapi/session.py

```diff
@@ -15,12 +15,54 @@
 #    License for the specific language governing permissions and limitations
 #    under the License.
 
+import abc
+import itertools
+import six
+
+from oslo_log import log as logging
+from oslo_utils import excutils
 from oslo_vmware import api
+from oslo_vmware import exceptions as vexc
 from oslo_vmware import vim
+from oslo_vmware.vim_util import get_moref_value
 
 import nova.conf
 
 CONF = nova.conf.CONF
+LOG = logging.getLogger(__name__)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class StableMoRefProxy(object):
+    """Abstract Basis class which acts as a proxy
+    for Managed-Object-References (MoRef).
+    Those references are usually "stable", meaning
+    they don't change over the life-time of the object.
+
+    But usually doesn't mean always. In that case, we
+    need to fetch the reference again via some search method,
+    which uses a guaranteed stable identifier (names, uuids, ...)
+    """
+    def __init__(self, ref):
+        self.moref = ref
+
+    @property
+    def __class__(self):
+        # Suds uses instanceof() to determine the serialiser,
+        # instead of duck-typing, so we have to fake it
+        return self.moref.__class__
+
+    @abc.abstractmethod
+    def fetch_moref(self, session):
+        """Updates the moref field or raises
+        same exception the initial search would have
+        """
+
+    def __getattr__(self, name):
+        return getattr(self.moref, name)
+
+    def __repr__(self):
+        return "{}({!r})".format(self.__class__.__name__, self.moref)
```

Comment:
If we would be able to overwrite `__class__` above, then this would show the class of `self.moref` and not ours. I think the intention was to show ours, right?

Reply:
True, I'll replace it simply with `StableMoRefProxy`, otherwise it will get quite complex to find out the proper subclass.